### DISCUSSION TITLE
GradCAM visualisation with pytorch-lightning

### DISCUSSION THREAD

**Discussion by ik362** [Project-MONAI/MONAI#2622] [Q&A]
Hi there,
I am using pytorch-lightning to create a custom model and I would like to use GradCAM to visualise the output of my self-attention layer.

So far I have not found any documantation/examples on how MONAI can be implemented within a pytorch-lightning module so I wanted to ask: how MONAI visualisations could be implemented in a pytorch-lightning module?

Thanks for any help/suggestions!

**Comment by wyli** [comment]
There is a thread might be useful: https://github.com/Project-MONAI/MONAI/discussions/1598

**Reply by ik362** [reply]
Hi @wyli,
Thanks for pointing me in that direction!

I was hoping you might be able to help me understanding the output. In this line `cam_img = self.cam(x = raw_img, class_idx = None)` my `raw_img` has dimensions `(193, 229, 193)`. However the output `cam_img` has dimensions `(1, 193, 229)`.

I am aming to get the `grad_img` for all the slices in my scan. Is this possible?

Thanks!

**Reply by rijobro** [reply]
You need to add both batch and channel dimensions to your data before passing it through GradCAM. Have you done that?

**Reply by ik362** [reply]
Hi @rijobro,
Sorry, I should have been more specific (my bad).

`self.cam` is defined as `self.cam = visualize.GradCAM(nn_module = model, target_layers = '4')`

And my test step is:
```python
def test_step(self, batch, batch_idx):
    raw_img, target = batch
    y_hat = self.forward(raw_img)
    acc = self.accuracy(y_hat, target)
    with torch.set_grad_enabled(True):
        print(raw_img.shape)
        cam_img = self.cam(x = raw_img, class_idx = None)
        print(cam_img.shape)
    self.log('test_acc', acc)
    return acc
```
The shape of raw_img is `([1, 193, 229, 193])` and cam_img is `([1, 1, 193, 229])`

Thanks!

**Reply by rijobro** [reply]
what happens when you do: `cam_img = self.cam(x = raw_img[None], class_idx = None)`?

**Reply by ik362** [reply]
Hi @rijobro,

When I change to `raw_img[None]` I get the runtime error:
```python
RuntimeError: Expected 4-dimensional input for 4-dimensional weight [193, 193, 3, 3], but got 5-dimensional input
of size [1, 1, 193, 229, 193] instead
```

**Reply by Nic-Ma** [reply]
Hi @ik362 ,

Could you please help provide the full test program for further debug?
As @rijobro suggested, the input data should have both `batch` and `channel` dims(for example: `[1, 1, 193, 229, 193]`), but I don't understand why you said the previous shape of `raw_img` is `[1, 193, 229, 193]`?

Thanks.

**Reply by ik362** [reply]
Hi @Nic-Ma,

Sorry for the delay in getting back to you. Is it the `pl.LightningModule` that you want provided or my entire notebook?

Thanks!

**Reply by Nic-Ma** [reply]
Please paste the full program to debug the data shape, I think the shape of `raw_img` may not be correct.
@rijobro What do you think?

Thanks.

**Reply by rijobro** [reply]
Hi @ik362 , it's hard to figure out why it's not working for you. It would be useful to create a minimal example showing what's going on. It can have the following structure:
```python
# some imports

# create a network

# create gradcam

# create random pytorch tensor

# perform gradcam
```
When you get that working, you can build up the complexity until you get back to your current use case.

If you haven't done so already, it might also be worth updating your MONAI version.

**Reply by ik362** [reply]
Hi @Nic-Ma and @rijobro,

I created a minimal example as a colab notebook: https://colab.research.google.com/drive/1DdqFkdFscZosFgXlbHxgGsuZMHDCTIag?usp=sharing

Thanks for having a look, much appreciated!

**Reply by rijobro** [reply]
Ignoring the GradCAM for a second, does your current code give you results that you expect?

The code is telling me that you're using 2D images with spatial dimensions of (229, 193), and that there are 193 channels. Is this correct? I just want to check that you aren't trying to create 3D images with spatial dimensions of (193, 229, 193).

**Reply by ik362** [reply]
Hi @rijobro,

Thanks for taking a look (you will have to forgive me I am a relative newbie), you are correct: my aim is to use an entire 3D T1w scan rather than a stack of 2D images. I changed the dummy data so that the data now has size`[1, 193, 229, 193]`, i.e. 1 channel and spatial dimensions (193, 229, 193). Is this correct?

Thanks!

**Reply by ik362** [reply]
Hi @rijobro,
Just wanted to give this discussion a bump to see if there might be a fix for the problem above?
Thanks!

**Reply by rijobro** [reply]
Hi, sorry you're still having the same problem. In answer to your previous question, the input to the GradCAM expects the input image to have the shape: `B, C, H, W, [D]`. So your dummy data should probably have the shape `[1, 1, 193, 229, 193]`, such that it has B=1, C=1, H=193, W=229, D=193.
